lib/IRGen/GenClangType.cpp

```diff
@@ -403,7 +403,19 @@ clang::CanQualType GenClangType::visitProtocolType(CanProtocolType type) {
 }
 
 clang::CanQualType GenClangType::visitMetatypeType(CanMetatypeType type) {
-  return getClangMetatypeType(getClangASTContext());
+  assert(type->hasRepresentation() &&
+         "metatype should have been assigned a representation by SIL");
+  switch (type->getRepresentation()) {
+  case MetatypeRepresentation::Thin:
+    return getClangASTContext().VoidTy;
+
+  case MetatypeRepresentation::Thick:
+    llvm_unreachable("thick metatypes don't have a corresponding Clang type");
+
+  case MetatypeRepresentation::ObjC:
+    return getClangMetatypeType(getClangASTContext());
+  }
+  llvm_unreachable("bad representation");
```

Comment:
For the changes in this file, please also update the corresponding code in `ClangTypeConverter.cpp`.

Reply:
Thanks for pointing this out -- I hadn't seen the comment at the top of `IRGen/GenClangType.cpp` saying to do this.

Two questions:

* In IRGen, I can rely on the metatype's representation being populated (see assertion that I added above, copied from `TypeConverter::convertMetatypeType()`). But can I also rely on the representation being populated in `AST/ClangTypeConverter.cpp`? If not, what should I do -- check whether we have a representation, and if we don't, just make an assumption...?

* Any insights into why there's this duplication between `AST/ClangTypeConverter.cpp` and `IRGen/GenClangType.cpp`? The main difference between them seems to be that the latter computes canonical types while the latter doesn't, but that could presumably be dealt with via a wrapper?

Reply:
> In IRGen, I can rely on the metatype's representation being populated (see assertion that I added above, copied from TypeConverter::convertMetatypeType()). But can I also rely on the representation being populated in AST/ClangTypeConverter.cpp?

I'm not 100% sure but I think you should be able to rely on it. If you see any failure purely because of this change, you can file a bug report on JIRA for investigation (and assign it to me) leaving a placeholder branch here to avoid a crash, say something like:

```
// FIXME: SR-NNNNN Unclear if we should expect to have a representation here.
if (!type->hasRepresentation()) {
    return getClangIdType(clangCtx);
}
```

Since the results of `ClangTypeConverter.cpp` are not really being used per se, it's ok if we have that here as a stop-gap until we start using it.

> Any insights into why there's this duplication between AST/ClangTypeConverter.cpp and IRGen/GenClangType.cpp? The main difference between them seems to be that the latter computes canonical types while the latter doesn't, but that could presumably be dealt with via a wrapper?

Going forward, we want to be able to track Clang types in the AST and SIL. (You'll see that `FunctionType` has a field for it and I have an open PR for SIL which I haven't had time to work on recently). The plan is to eventually get rid of the IRGen one and use the AST one.

There are some subtle differences the two implementations are not exactly the same, but I _think_ it shouldn't matter for this change.

Reply:
OK, thanks. Done!

(Finally -- sorry, I was busy with other things and it took me a while to get back to this.)